src/combinatorics/hales_jewett.lean

```diff
@@ -0,0 +1,299 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import data.fintype.basic
+import algebra.big_operators.basic
+import data.equiv.fin
+
+/-!
+# The Hales-Jewett theorem
+
+We prove the Hales-Jewett theorem and deduce Van der Waerden's theorem as a corollary.
+
+We consistently work with finite types instead of natural numbers. That is, we write `m, n, k` for
+(finite) types where one might traditionally take `m, n, k` to be natural numbers. This allows us to
+work directly with `m`, `option m`, `(n → m) → k`, and `n ⊕ n'` instead of `fin m`, `fin (m+1)`,
+`fin (k^(m^n))`, and `fin (n + n')`.
+
+We take '`X` is monochromatic' to mean 'there exists a color `c` such that every point of `X` has
+the color `c`' as opposed to 'any `x y : X` have the same color' since the former is generally
+easier to work with. This means that our formulation of Hales-Jewett would be false when `m` and `k`
+are both empty (then there can be a coloring even though there is no color). Thus we require `m` to
+be nonempty in the statement of `exists_mono_in_high_dimension`.
```

Comment:
Upon further reflection this is false... when `m = k = n = empty` there can be no `empty`-colouring of `empty -> empty`. I'll try to get rid of the `[nonempty m]` assumption in H-J eventually.